pretixprint/app/src/main/java/eu/pretix/pretixprint/ui/USBSettingsFragment.kt

```diff
@@ -14,34 +14,47 @@ import android.view.LayoutInflater
 import android.view.View
 import android.view.ViewGroup
 import android.widget.Button
+import android.widget.CheckBox
 import androidx.annotation.RequiresApi
 import androidx.preference.PreferenceManager
 import com.google.android.material.dialog.MaterialAlertDialogBuilder
-import com.google.android.material.switchmaterial.SwitchMaterial
 import com.google.android.material.textfield.TextInputEditText
+import com.google.android.material.textfield.TextInputLayout
 import eu.pretix.pretixprint.R
 import splitties.toast.toast
 
 class USBSettingsFragment : SetupFragment() {
-    private val ACTION_USB_PERMISSION = "eu.pretix.pretixprint.settings.USB_PERMISSION"
+    companion object {
+        private const val ACTION_USB_PERMISSION = "eu.pretix.pretixprint.settings.USB_PERMISSION"
+    }
 
     private val usbReceiver = object : BroadcastReceiver() {
         @RequiresApi(Build.VERSION_CODES.LOLLIPOP)
         override fun onReceive(context: Context, intent: Intent) {
-            if (ACTION_USB_PERMISSION == intent.action) {
-                synchronized(this) {
-                    val device: UsbDevice? = intent.getParcelableExtra(UsbManager.EXTRA_DEVICE)
+            if (ACTION_USB_PERMISSION != intent.action) {
+                return
+            }
 
-                    if (intent.getBooleanExtra(UsbManager.EXTRA_PERMISSION_GRANTED, false)) {
-                        if (device!!.serialNumber != null && device!!.serialNumber != "null") {
-                            view?.findViewById<TextInputEditText>(R.id.teSerial)?.setText(device!!.serialNumber)
-                        } else {
-                            view?.findViewById<TextInputEditText>(R.id.teSerial)?.setText("${Integer.toHexString(device!!.vendorId)}:${Integer.toHexString(device!!.productId)}")
+            synchronized(this) {
+                if (!intent.getBooleanExtra(UsbManager.EXTRA_PERMISSION_GRANTED, false)) {
+                    toast(R.string.err_usb_permission_denied)
+                    return
+                }
+
+                val device: UsbDevice = intent.getParcelableExtra(UsbManager.EXTRA_DEVICE) ?: return
+                val deviceId = "${String.format("%04x", device.vendorId)}:${String.format("%04x", device.productId)}"
+                if (device.serialNumber != null && device.serialNumber != "null") {
+                    view?.findViewById<TextInputLayout>(R.id.tilSerial)?.apply {
+                        endIconMode = TextInputLayout.END_ICON_CUSTOM
+                        isEndIconVisible = true
```

Comment:
If I now click "select device" again and choose one without a serial number, I think we are missing a cleanup of the icon (have not tested it)